### PR TITLE
[Dependency]add switch for libaio

### LINKED ISSUE #45: [Dependency]Make libaio optional
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/async_simple/issues) and found no similar issues.

### What happened + What you expected to happen
libaio just used in example, so it can be optional to minimal dependency.

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by qicosmos**
## Why

Close #45 

## What is changing
Add ENABLE_LIBAIO flag for  libaio.

If "SET(ENABLE_LIBAIO ON)" will require libaio, otherwise no need.

### CODE REVIEW

File: async_simple/executors/test/SimpleIOExecutorTest.cpp [hunk 1]
```diff
@@ -51,21 +51,21 @@ class SimpleIOExecutorTest : public FUTURE_TESTBASE {
...
-    auto fd = open(kTestFile, O_RDWR | O_DIRECT | O_CREAT, 0600);
+    auto fd = open(kTestFile, O_RDWR | O_DIRECT | O_CREAT);
     auto output = memalign(4096, kBufferSize);
     memcpy((char*)output, expect.data(), expect.length());
     _executor->submitIO(
         fd, IOCB_CMD_PWRITE, output, expect.length(), 0,
-        [](io_event& event) mutable { EXPECT_EQ(4096, (int32_t)event.res); });
-    std::this_thread::sleep_for(300ms);
+        [](io_event_t& event) mutable { EXPECT_EQ(4096, (int32_t)event.res); });
+    usleep(1000 * 300);
```
Comment (ChuanqiXu9) [inline async_simple/executors/test/SimpleIOExecutorTest.cpp:60]: This change is not intentional, right?

File: CMakeLists.txt [hunk 2]
```diff
@@ -5,16 +5,20 @@ enable_testing()
 list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
 message(STATUS "CMAKE_MODULE_PATH: ${CMAKE_MODULE_PATH}")
 
-find_package(Aio REQUIRED)
```
Comment (ChuanqiXu9) [inline CMakeLists.txt:8]: I would like remain `find_package` here. We just don't compile things require libaio if we couldn't find one. This is more friendly to users. We could emit a macro in CMakeLists to detect if we find libaio.

File: async_simple/executors/test/SimpleIOExecutorTest.cpp [hunk 3]
```diff
@@ -51,21 +51,21 @@ class SimpleIOExecutorTest : public FUTURE_TESTBASE {
 
 TEST_F(SimpleIOExecutorTest, testNormal) {
     std::string expect(4096, '0');
-    auto fd = open(kTestFile, O_RDWR | O_DIRECT | O_CREAT, 0600);
+    auto fd = open(kTestFile, O_RDWR | O_DIRECT | O_CREAT);
```
Comment (ChuanqiXu9) [inline async_simple/executors/test/SimpleIOExecutorTest.cpp:54]: Is it necessary?

File: async_simple/executors/test/SimpleIOExecutorTest.cpp (hunk 1 above)
Comment (qicosmos) [reply async_simple/executors/test/SimpleIOExecutorTest.cpp:60]: Maybe it's from an old version, i will confirm.

File: async_simple/executors/test/SimpleIOExecutorTest.cpp (hunk 3 above)
Comment (qicosmos) [reply async_simple/executors/test/SimpleIOExecutorTest.cpp:54]: Maybe the code is from an old version, i will confirm.

File: CMakeLists.txt (hunk 2 above)
Comment (qicosmos) [reply CMakeLists.txt:8]: OK

File: CMakeLists.txt [hunk 4]
```diff
@@ -10,6 +10,8 @@ find_package(GMock REQUIRED)
 find_package(GTest REQUIRED)
 find_package(Benchmark)
 
+SET(ENABLE_LIBAIO OFF)
```
Comment (ChuanqiXu9) [inline CMakeLists.txt:13]: We don't need a variable controlled by user to indicate whether the LIBAIO is enabled. We could do it automatically. I mean, if we found libaio, build the codes related with libaio. If we couldn't find libaio, then just skip the codes depends libaio. We could refer to the implementation of `Benchmark`. We need to emit an macro too.

File: async_simple/executors/test/SimpleIOExecutorTest.cpp [hunk 5]
```diff
@@ -76,8 +76,8 @@ TEST_F(SimpleIOExecutorTest, testException) {
     memset(output, 0, kBufferSize);
     _executor->submitIO(
         -1, IOCB_CMD_PREAD, output, kBufferSize, 0,
-        [](io_event& event) mutable { EXPECT_TRUE((int32_t)event.res < 0); });
-    std::this_thread::sleep_for(300ms);
+        [](io_event_t& event) mutable { EXPECT_TRUE((int32_t)event.res < 0); });
+    usleep(1000 * 300);
```
Comment (ChuanqiXu9) [inline async_simple/executors/test/SimpleIOExecutorTest.cpp:80]: This is not intentional.

File: CMakeLists.txt [hunk 6]
```diff
@@ -10,6 +10,8 @@ find_package(GMock REQUIRED)
...
+
 if(LIBAIO_INCLUDE_DIR AND LIBAIO_LIBRARIES)
```
Comment (ChuanqiXu9) [inline CMakeLists.txt:15]: This would still emit an error if we don't find one.

**Review by ChuanqiXu9 — CHANGES_REQUESTED** [review]

**Comment by RainMark** [comment]
why not delete the libaio directly?

**Comment by qicosmos** [comment]
> why not delete the libaio directly? Use another thread pool to process io request

I want to make libaio optional firstly, i think removing it can be in another PR.
And we need more discussion and design about the new IOExecutor interface.

**Comment by ChuanqiXu9** [comment]
> why not delete the libaio directly? Use another thread pool to process io request

According to our discussion, we should refactor the interface for IOExecutor and remain basic read/write API in async_simple. And add other high level IOExecutor like AIOExecutor, IOUringExecutor and ASIOExecutor in another directory which is parallel with the main dir of async_simple. So if we are not going to delete it in the end, I think we'd better keep it as possible.

File: CMakeLists.txt [hunk 7]
```diff
@@ -5,16 +5,18 @@ enable_testing()
 list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
```
Comment (ChuanqiXu9) [inline CMakeLists.txt:5]: We should update the minimum_require to 3.12 since `add_compile_definitions ` is introduced in 3.12: https://cmake.org/cmake/help/latest/command/add_compile_definitions.html

File: CMakeLists.txt [hunk 8]
```diff
@@ -62,12 +64,16 @@ set(HEADERS_PATH
     ${PROJECT_SOURCE_DIR}
     ${GTEST_INCLUDE_DIR}
     ${GMOCK_INCLUDE_DIR}
-    ${AIO_INCLUDE_DIR}
+    if (ENABLE_LIBAIO)
+        ${AIO_INCLUDE_DIR}
+    endif()
```
Comment (ChuanqiXu9) [inline CMakeLists.txt:69]: Unnecessary change.

File: CMakeLists.txt [hunk 9]
```diff
@@ -62,12 +64,16 @@ set(HEADERS_PATH
...
     )
 include_directories(${HEADERS_PATH})
 
 set(deplibs)
-list(APPEND deplibs ${LIBAIO_LIBRARIES})
+if (ENABLE_LIBAIO)
+    list(APPEND deplibs ${LIBAIO_LIBRARIES})
+endif()
 list(APPEND deplibs pthread)
```
Comment (ChuanqiXu9) [inline CMakeLists.txt:77]: Unnecessary change.

File: README.md [hunk 10]
```diff
@@ -21,7 +21,9 @@ and the traditional Future/Promise.
 ## Using apt (ubuntu and debian's)
 
 ```
+// note: if "SET(ENABLE_LIBAIO ON)" will require libaio, otherwise no need.
```
Comment (ChuanqiXu9) [inline README.md:24]:
```suggestion
# Optional.
```

File: README.md [hunk 11]
```diff
@@ -21,7 +21,9 @@ and the traditional Future/Promise.
...
 sudo apt install libaio-dev -y
+
```
Comment (ChuanqiXu9) [inline README.md:26]: Unnecessary change

File: README.md [hunk 12]
```diff
@@ -37,14 +39,17 @@ cd .. && sudo rm -rf build
 
 ## Using yum (CentOS and fedora)
 ```
+// note: if "SET(ENABLE_LIBAIO ON)" will require libaio, otherwise no need.
```
Comment (ChuanqiXu9) [inline README.md:42]:
```suggestion
# Optional.
```

File: README_CN.md [hunk 13]
```diff
@@ -15,14 +15,15 @@ async\_simple是阿里巴巴开源的轻量级C++异步框架。提供了基于C
 
 # 准备环境
 
-async\_simple涉及C++20协程，对编译器版本有较高要求。需要clang11，gcc10及其以上版本。 其他只依赖gtest和libaio。
+async\_simple涉及C++20协程，对编译器版本有较高要求。需要clang11，gcc10及其以上版本。 其他只依赖gtest，可选依赖libaio。
```
Comment (ChuanqiXu9) [inline README_CN.md:18]:
```suggestion
async\_simple 涉及 C++20 协程，对编译器版本有较高要求。需要 clang11 或 gcc10 及其以上版本。编译运行测试需要 gtest。需要 libaio 的功能的需要依赖 libaio。
```

File: README_CN.md [hunk 14]
```diff
@@ -15,14 +15,15 @@ async\_simple是阿里巴巴开源的轻量级C++异步框架。提供了基于C
...
 
 ## Debian/Ubuntu系列
 
 - 安装clang11及其以上版本。安装方法见：[APT Packages](https://apt.llvm.org/)
-- 安装cmake、libaio。
+- 安装cmake、libaio(可选)。
```
Comment (ChuanqiXu9) [inline README_CN.md:23]:
```suggestion
- 安装cmake。
- 如需使用 libaio 功能，需安装 libaio。
```

File: README_CN.md [hunk 15]
```diff
@@ -15,14 +15,15 @@ async\_simple是阿里巴巴开源的轻量级C++异步框架。提供了基于C
...
 
 ```bash
+// 如果设置"SET(ENABLE_LIBAIO ON)"将需要依赖libaio, 否则不依赖libaio.
```
Comment (ChuanqiXu9) [inline README_CN.md:26]:
```suggestion
# Optional.
```

File: README.md [hunk 16]
```diff
@@ -37,14 +39,17 @@ cd .. && sudo rm -rf build
...
+
 sudo yum install cmake -y
 # Try to build gtest and gmock from source
 ```
 
 ## Build Dependencies From Source
 
 ```
+// note: if "SET(ENABLE_LIBAIO ON)" will require libaio, otherwise no need.
 # libaio
```
Comment (ChuanqiXu9) [inline README.md:53]:
```suggestion
# libaio (optional)
```

File: async_simple/coro/test/CMakeLists.txt [hunk 17]
```diff
@@ -1,6 +1,8 @@
...
+if (ENABLE_LIBAIO)
+    file(GLOB coro_test_src "*.cpp")
+    add_executable(async_simple_coro_test ${coro_test_src} ${PROJECT_SOURCE_DIR}/async_simple/test/dotest.cpp)
 
-target_link_libraries(async_simple_coro_test async_simple ${deplibs} ${testdeplibs})
+    target_link_libraries(async_simple_coro_test async_simple ${deplibs} ${testdeplibs})
 
-add_test(NAME run_async_simple_coro_test COMMAND async_simple_coro_test)
+    add_test(NAME run_async_simple_coro_test COMMAND async_simple_coro_test)
+endif()
```
Comment (ChuanqiXu9) [inline async_simple/coro/test/CMakeLists.txt:8]: Unnecessary change.

File: async_simple/executors/SimpleExecutor.h [hunk 18]
```diff
@@ -19,7 +19,6 @@
 #include <functional>
 
 #include <async_simple/Executor.h>
-#include <async_simple/executors/SimpleIOExecutor.h>
```
Comment (ChuanqiXu9) [inline async_simple/executors/SimpleExecutor.h:22]: Unnecessary change.

File: async_simple/executors/SimpleExecutor.h [hunk 19]
```diff
@@ -51,9 +50,7 @@ class SimpleExecutor : public Executor {
     SimpleExecutor(size_t threadNum) : _pool(threadNum) {
         [[maybe_unused]] auto ret = _pool.start();
         assert(ret);
-        _ioExecutor.init();
```
Comment (ChuanqiXu9) [inline async_simple/executors/SimpleExecutor.h:54]: Unnecessary change.

File: async_simple/executors/SimpleExecutor.h [hunk 20]
```diff
@@ -51,9 +50,7 @@ class SimpleExecutor : public Executor {
...
     }
-    ~SimpleExecutor() { _ioExecutor.destroy(); }
```
Comment (ChuanqiXu9) [inline async_simple/executors/SimpleExecutor.h:56]: Unnecessary change.

File: async_simple/executors/SimpleExecutor.h [hunk 21]
```diff
@@ -87,11 +84,8 @@ class SimpleExecutor : public Executor {
                util::ThreadPool::ERROR_NONE;
     }
 
-    IOExecutor* getIOExecutor() override { return &_ioExecutor; }
```
Comment (ChuanqiXu9) [inline async_simple/executors/SimpleExecutor.h:90]: Unnecessary change. This patch shouldn't break the functionality.

File: async_simple/executors/SimpleExecutor.h [hunk 22]
```diff
@@ -87,11 +84,8 @@ class SimpleExecutor : public Executor {
...
-
 private:
     util::ThreadPool _pool;
-    SimpleIOExecutor _ioExecutor;
```
Comment (ChuanqiXu9) [inline async_simple/executors/SimpleExecutor.h:94]: Unnecessary change.

File: async_simple/executors/test/CMakeLists.txt [hunk 23]
```diff
@@ -1,6 +1,8 @@
...
+if (ENABLE_LIBAIO)
+    file(GLOB executor_test_src "*.cpp")
+    add_executable(async_simple_executor_test ${executor_test_src} ${PROJECT_SOURCE_DIR}/async_simple/test/dotest.cpp)
 
-target_link_libraries(async_simple_executor_test async_simple ${deplibs} ${testdeplibs})
+    target_link_libraries(async_simple_executor_test async_simple ${deplibs} ${testdeplibs})
 
-add_test(NAME run_async_simple_executor_test COMMAND async_simple_executor_test)
+    add_test(NAME run_async_simple_executor_test COMMAND async_simple_executor_test)
+endif()
```
Comment (ChuanqiXu9) [inline async_simple/executors/test/CMakeLists.txt:8]: Unnecessary change.

**Review by ChuanqiXu9 — COMMENTED** [review]
We should use macro directive like:
```
#ifdef ASYNC_SIMPLE_HAS_NOT_AIO
#ifndef  ASYNC_SIMPLE_HAS_NOT_AIO
```

in SimpleExecutor.h and SimpleIOExecutor.cpp to wrap the things which is needed by libaio.

**Comment by RainMark** [comment]
> > why not delete the libaio directly? Use another thread pool to process io request
> 
> According to our discussion, we should refactor the interface for IOExecutor and remain basic read/write API in async_simple. And add other high level IOExecutor like AIOExecutor, IOUringExecutor and ASIOExecutor in another directory which is parallel with the main dir of async_simple. So if we are not going to delete it in the end, I think we'd better keep it as possible.

ok

**Comment by RainMark** [comment]
> > why not delete the libaio directly? Use another thread pool to process io request
> 
> I want to make libaio optional firstly, i think removing it can be in another PR. And we need more discussion and design about the new IOExecutor interface.

👌🏻

**Comment by ChuanqiXu9** [comment]
Have you tested this under an environment without libaio?

**Comment by qicosmos** [comment]
> 

I remove libaio to test, it passed.

**Review by RainMark — APPROVED** [review]

**Comment by RainMark** [comment]
LGTM

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM.
